x-pack/platform/plugins/shared/fleet/server/tasks/automatic_agent_upgrade_task.ts

```diff
@@ -0,0 +1,308 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { SavedObjectsClient } from '@kbn/core/server';
+import type {
+  CoreSetup,
+  ElasticsearchClient,
+  Logger,
+  SavedObjectsClientContract,
+} from '@kbn/core/server';
+import type {
+  ConcreteTaskInstance,
+  TaskManagerSetupContract,
+  TaskManagerStartContract,
+} from '@kbn/task-manager-plugin/server';
+import { getDeleteTaskRunResult } from '@kbn/task-manager-plugin/server/task';
+import type { LoggerFactory } from '@kbn/core/server';
+import { errors } from '@elastic/elasticsearch';
+import semverGt from 'semver/functions/gt';
+
+import type { Agent, FleetServerAgentMetadata } from '../../common/types';
+
+import { agentPolicyService, appContextService } from '../services';
+import {
+  fetchAllAgentsByKuery,
+  getTotalAgentsByKuery,
+  sendUpgradeAgentsActions,
+} from '../services/agents';
+import { AGENT_POLICY_SAVED_OBJECT_TYPE } from '../constants';
+import { AgentStatusKueryHelper, isAgentUpgradeable } from '../../common/services';
+
+export const TYPE = 'fleet:automatic-agent-upgrade-task';
+export const VERSION = '1.0.0';
+const TITLE = 'Fleet Automatic agent upgrades';
+const SCOPE = ['fleet'];
+const INTERVAL = '1h'; // TODO: check
+const TIMEOUT = '10m'; // // TODO: check
+const AGENTS_BATCHSIZE = 10000; // // TODO: check
+const AGENT_POLICIES_BATCHSIZE = 500; // // TODO: check
```

Comment:
we should add tests that multiple batches of agents and agent policies are upgraded correctly